glslc/src/dependency_info.h

```diff
@@ -0,0 +1,163 @@
+  // file names saved in container 'dependent_files' to dependency info file or
+  // compilation output string according to its dumping mode.
+  //
+  // When the handler is set to dump dependency info as extra dependency info
+  // files, this method will open the files with deduced names and write the
+  // dependency info to them. Error messages caused by writing the files are
+  // emitted to stderr.
+  //
+  // When the handler is set to dump dependency info as compilation output, the
+  // compilation output string, which is passed through compilation_output_ptr,
+  // will be cleared and this method will write dependency info to it. Then the
+  // dependency info should be emitted as normal compilation output.
+  //
+  // If the dump mode is not set when this method is called, return false.
+  bool DumpDependencyInfo(std::string compilation_output_file_name,
+                          std::string source_file_name,
+                          std::string* compilation_output_ptr,
+                          const std::set<std::string>& dependent_files);
+
+  // Sets to always dump dependency info as an extra file, instead of the normal
+  // compilation output. This means the output name specified by -o options
+  // won't be used for the dependency info file.
+  void SetDumpToExtraDependencyInfoFiles() { mode_ = dump_as_extra_file; };
+
+  // Sets to dump dependency info as normal compilation output. The dependency
+  // info will be either saved in a file with -o option specified file, or, if
+  // no output file name specified, to stdout.
+  void SetDumpAsNormalCompilationOutput() {
+    mode_ = dump_as_compilation_output;
+  }
+
+  // Checks if the handler's dumping mode is set to dump dependency info as
+  // extra dependency info files.
+  bool DumpingToExtraDependencyInfoFiles() {
+    return mode_ == dump_as_extra_file;
+  }
+
+  // Checks if the handler's dumping mode is set to dump dependency info as
+  // normal compilation output.
+  bool DumpingAsCompilationOutput() {
+    return mode_ == dump_as_compilation_output;
+  }
+
+  // Checks if the handler's dumping mode is not set.
+  bool DumpingModeNotSet() { return mode_ == not_set; }
+
+  // Checks if the handler is at valid state for dumping dependency info.
+  bool IsValid(std::string* error_msg_ptr, size_t num_files) {
+    if (DumpingModeNotSet()) {
+      *error_msg_ptr =
+          "to generate dependencies you must specify either -M (-MM) or -MD";
+      return false;
+    }
+
+    if (!user_specified_dep_file_name_.empty() ||
+        !user_specified_dep_target_label_.empty()) {
+      if (num_files > 1) {
+        *error_msg_ptr =
+            "to specify dependency info file name or dependency info target, "
+            "only one input file is allowed.";
+        return false;
+      }
+    }
+    return true;
+  }
+
+ private:
+  typedef enum {
+    // By default, dump mode should be set to not_set, which means the
```

Comment:
This is too much information for the definition of this enum value.
Youd can just say:  not_set means don't generate dependency info.
How and when it's *USED* is another matter, dealt with elsewhere.